specs/OAuth/AuthManager.md

```diff
@@ -0,0 +1,891 @@
+AuthManager API
+===
+
+This is the spec for proposal: [Issue #441](https://github.com/microsoft/WindowsAppSDK/issues/441)
+
+This spec details the API of a new `AuthManager` in WinAppSDK,
+enabling desktop applications such as WinUI3 to seamlessly perform OAuth functionality across diverse Windows platforms.
+
+
+
+# Background
+
+The current OS 
+[WebAuthenticationBroker](https://docs.microsoft.com/uwp/api/windows.security.authentication.web.webauthenticationbroker),
+primarily designed for UWP applications,
+presents several challenges when used in desktop environments. 
+Key issues include the dependency on `ApplicationView`, 
+which are not compatible with desktop app frameworks.
+As a result, developers are forced to resort to workarounds involving interop interfaces and 
+additional code to implement OAuth 2.0 functionality into their WinUI 3 desktop applications.
+
+In response to this feedback, this `AuthManager` API for WinAppSDK aims to provide 
+a streamlined solution that meets the expectations of our developer community. 
+The goal is to offer seamless OAuth 2.0 capabilities with full feature parity across all Windows platforms supported by WinAppSDK. 
+This new API eliminates the need for cumbersome workarounds and 
+simplifies the process of incorporating OAuth 2.0 functionality into desktop applications.
+
+
+This new WinAppSDK version of the `AuthManager` is different than the existing public
+[WebAuthenticationBroker](https://learn.microsoft.com/en-us/uwp/api/windows.security.authentication.web.webauthenticationbroker?view=winrt-22621) 
+API. We've opted to follow OAuth best practices more closely - e.g. using user's default browser. The best practices for the API are taken from the
+IETF(Internet Engineering Task Force) OAuth 2.0 Authorization Framework [RFC 6749](https://tools.ietf.org/html/rfc6749), 
+Proof Key for Code Exchange (PKCE) [RFC 7636](https://tools.ietf.org/html/rfc7636),
+and OAuth 2.0 for Native Apps [RFC 8252](https://tools.ietf.org/html/rfc8252).
+
+
+
+# Conceptual pages
+
+ ## Perform OAuth 2.0 (c#)
+
+ Performing an Authorization Code Request (grant type/'response_type' = "code")
+
+ ```c#
+AuthRequestParams authRequestParams = AuthRequestParams::CreateForAuthorizationCodeRequest(L"my_client_id",
+    Uri(L"my-app:/oauth-callback/"));
+authRequestParams.Scope(L"user:email user:birthday");
```

Comment:
Is the Scope setter needed? Usually the scope is part of the request URL that you get from the service, and not something you set separately. I wonder if this class is really needed? You just need the auth url and the callbackurl which could both be provided in `InitiateAuthRequestAsync`.

Reply:
Agreed, leaving more flexibility here and trying not to pin things down too specifically would be best as there's inevitably cases where the spec isn't implemented correctly that can cause problems with these scenarios.  As @dotMorten suggests, I think simplifying the `InitiateAuthRequestAsync` call to accept two uri parameters is a good approach.

Reply:
@dotMorten and @Redth 
Thanks for reviewing the spec :)
Another thought, 
I was thinking to keep the existing APIs and add an extra API which accepts two uri params, something like "InitiateAuthRequestAsync(uri,uri)". This gives user the flexibility to choose is they want to set the AuthRequestParams or go with the default implementation. 

If I understand correctly, accepting two URIs would involve an implementation in which the response_type ='token' and it would be an implicit authorization grant.
https://www.rfc-editor.org/rfc/rfc6749.html#page-8

Reply:
I do wonder if we can get this down to just a single step, like `OAuth2Token::RequestAsync("my_client_id", "https://my/server", "my-app:/oauth-callback")`.

Maybe that even implies the right naming convention ... it's an `OAuth2Token` and the static verb action is `Request-` which is what we use for "things that might need to interact with the user."

Reply:
For simplicity, we can have one API like InitiateRequestAsync( "https://my/server/authorize?..") 

Here, we expect this uri ("https://my/server/authorize?..") to be a complete uri with the query parameters as below.
"https://<providerendpoint>?client_id=<clientid>&scope=<scopes>&response_type=token&redirect_uri=my-app:/oauth-callback"

As part of processing in the API, we can add the state parameter to it. 

@jonwis @Redth @AjitSurana @dotMorten

Reply:
Also, we are planning to keep the existing APIs and add this extra one for simplicity and it will support response_type="token". 
We want to keep the existing ones as well, in case, someone wants to use the response_type="code" for their OAuth.

Reply:
Based on further discussions, 
We are adding two new APIs for simplication.
1. InitiateRequestAsync(Uri fullAuthorizeURI, Uri RedirectURI)
2. InitiateRequestAsyncWithoutCallback(Uri fullAuthorizeUri)

We will provide examples for these two APIs at the top of spec doc.
These two APIs will support response_type="token".

CC: @jonwis @dotMorten @Redth @AjitSurana